pkg/scheduler/framework/preemption/preemption.go

```diff
@@ -401,6 +460,71 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
 	return nil
 }
 
+// prepareCandidateAsync triggers a goroutine for some preparation work:
+// - Evict the victim pods
+// - Reject the victim pods if they are in waitingPod map
+// - Clear the low-priority pods' nominatedNodeName status if needed
+// The Pod won't be retried until the goroutine triggered here completes.
+//
+// See http://kep.k8s.io/4832 for how the async preemption works.
+func (ev *Evaluator) prepareCandidateAsync(c Candidate, pod *v1.Pod, pluginName string) {
+	metrics.PreemptionVictims.Observe(float64(len(c.Victims().Pods)))
+
+	// Intentionally create a new context, not using a ctx from the scheduling cycle, to create ctx,
+	// because this process could continue even after this scheduling cycle finishes.
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := parallelize.NewErrorChannel()
+	preemptPod := func(index int) {
+		victim := c.Victims().Pods[index]
+		if err := ev.PreemptPod(ctx, c, pod, victim, pluginName); err != nil {
+			errCh.SendErrorWithCancel(err, cancel)
+		}
+	}
+
+	ev.mu.Lock()
+	ev.preempting.Insert(pod.UID)
+	ev.mu.Unlock()
+
+	logger := klog.FromContext(ctx)
+	go func() {
+		startTime := time.Now()
+		result := metrics.GoroutineResultSuccess
+		defer metrics.PreemptionGoroutinesDuration.WithLabelValues(result).Observe(metrics.SinceInSeconds(startTime))
+		defer metrics.PreemptionGoroutinesExecutionTotal.WithLabelValues(result).Inc()
+		defer func() {
+			ev.mu.Lock()
+			ev.preempting.Delete(pod.UID)
+			ev.mu.Unlock()
+			ev.Handler.Activate(logger, map[string]*v1.Pod{pod.Name: pod})
```

Comment:
why are we calling Activate? IIRC, Activate was previously used by Permit plugins, right?

This seems like an overload to Activate.

IIRC, when a scheduling cycle has a failure, we reschedule the pod directly (skipping unschedulable). IIUC, we want to do roughly the same here.

Reply:
So, the initial motivation is to address the failure case where the API call is failure at some points (https://github.com/kubernetes/kubernetes/pull/128170#discussion_r1827707578)

e.g., If the preemption API calls aren't successful at all, the preemptor Pod gets stuck in the unschedulable pod pool because no Pod/Delete event happens.
So, we have to somehow let the preemption plugin move the Pod out of the unsched pod pool in the failure case. That's why `Activate()` is a must here.

So, initially we ran `Activate()` only when the goroutine fails.
And, later we changed to always run `Activate()` after the preemption goroutine is done regardless of the goroutine result
because it can simplify the implementation and also because I _thought_ the goroutine is done == the pod is likely now schedulable.

That's all the context of the past discussion in this PR how the current `Activate()` forms.

_BUT_, rethinking about it, looks like we made a mistake. `Activate()` is a must for the failure scenario, that's correct. 
The incorrect part is the second part; the goroutine is done **!=** the pod is likely now schedulable.
The completion of the deletion APIs only means the victims just have `deletionTimestamp`, but does not mean the Pods are deleted.
So, let's say each victim takes 30 seconds (GracePeriod) to be completely deleted, the preemptor pod won't be schedulable for 30 seconds (* at least at the preemption target node).
If we `Activate()` the preemptor in that case, the preemptor's scheduling is retried and it consumes the scheduling cycle in vain.

Let me bring back the past implementation and make it run `Activate()` only when the goroutine fails

Reply:
cc @macsko @dom4ha 
(who joins the previous discussion so far

Reply:
> So, the initial motivation is to address the failure case where the API call is failure at some points 

This doesn't necessarily answer why we should reuse `Activate`. My pet-peeve is that this leads to adding the extra event with rare semantics and checks.

Could we just requeue the preemptor pod directly?

Reply:
Regarding the second part, my intention was to let other gating plugin to take over the gate keeping :) Apparently, there's no such functionality, although it could, as there should be still fit plugin saying no.

Reply:
This is how we do it in `handleBindingCycleError` https://github.com/kubernetes/kubernetes/blob/62bdf5d87c1242ef5894a51f3d79e51ebcd2a07d/pkg/scheduler/schedule_one.go#L345

and this is how we do it in `handleSchedulingFailure` https://github.com/kubernetes/kubernetes/blob/62bdf5d87c1242ef5894a51f3d79e51ebcd2a07d/pkg/scheduler/schedule_one.go#L1078
Since the pod wouldn't have any failed plugins (because it was an error), it would be directly put in back off here https://github.com/kubernetes/kubernetes/blob/62bdf5d87c1242ef5894a51f3d79e51ebcd2a07d/pkg/scheduler/backend/queue/scheduling_queue.go#L705
(I'm not sure how we handled this in scheduling hints)

Reply:
Ok, that sounds reasonable. I still don't love the approach, but there doesn't seem to be a reasonable alternative.

Please add the conclusions of this thread to the description.